torchani/extension/aev.cu

```diff
@@ -0,0 +1,627 @@
+
+
+    // Extract Rijs that is needed for RadialAEV comptuation i.e. all the Rij <= Rcr
+    int nRadialRij = cubDeviceSelect(d_Rij, d_radialRij, total_natom_pairs, d_count_out,
+				     [=]__device__(const PairDist<float> d){
+				       return d.Rij <= Rcr;
+				     });
+
+    int nblocks = (nRadialRij * 8 + block_size - 1)/block_size;
+    cuRadialAEVs<int, float, 8>
+      <<<nblocks, block_size>>>
+      (species_t.packed_accessor32<int,2,torch::RestrictPtrTraits>(),
+       ShfR_t.packed_accessor32<float,1,torch::RestrictPtrTraits>(),
+       EtaR_t.packed_accessor32<float,1,torch::RestrictPtrTraits>(),
+       aev_t.packed_accessor32<float,3,torch::RestrictPtrTraits>(),
+       d_radialRij,
+       aev_params,
+       nRadialRij);
+
+    // reuse buffer allocated for all Rij
+    // d_angularRij will store all the Rij required in Angular AEV computation
+    PairDist<float> *d_angularRij = d_Rij;
+    
+    // Extract Rijs that is needed for AngularAEV comptuation i.e. all the Rij <= Rca
+    int nAngularRij = cubDeviceSelect(d_radialRij, d_angularRij, nRadialRij, d_count_out,
+				      [=]__device__(const PairDist<float> d){
+					return d.Rij <= Rca;
+				      });
+
+    PairDist<float> *d_centralAtom = 0;
+    CubDebugExit(g_allocator.DeviceAllocate((void**)&d_centralAtom, sizeof(PairDist<float>) * nAngularRij));
+    //cudaMalloc((void **)&d_centralAtom, sizeof(PairDist<float>) * nAngularRij);
+    
+    
+    int *d_numPairsPerCenterAtom = 0;
+    CubDebugExit(g_allocator.DeviceAllocate((void **)&d_numPairsPerCenterAtom, sizeof(int) * nAngularRij));
+    //cudaMalloc((void **)&d_numPairsPerCenterAtom, sizeof(int) * nAngularRij);
+    
+    // group by center atom
+    int ncenter_atoms = cubEncode(d_angularRij, d_centralAtom, d_numPairsPerCenterAtom, nAngularRij, d_count_out);
+
+    int *d_centerAtomStartIdx = 0;
+    CubDebugExit(g_allocator.DeviceAllocate((void **)&d_centerAtomStartIdx, sizeof(int) * ncenter_atoms));
+    //cudaMalloc((void **)&d_centerAtomStartIdx, sizeof(int) * ncenter_atoms);
+    
+    cubScan(d_numPairsPerCenterAtom, d_centerAtomStartIdx, ncenter_atoms);    
+    {
+
+      //ncenter_atoms = 1;
+      const int nthreads_per_catom = 32;
+      const int nblocks_angAEV = (ncenter_atoms * nthreads_per_catom + block_size - 1)/block_size;
+      auto smem_size = [&aev_params](int max_nbrs, int ncatom_per_tpb){
+	int sm_aev = sizeof(float) * align<4>(aev_params.angular_length);
+	int sxyz = sizeof(float) * max_nbrs * 3;
+	int sRij = sizeof(float) * max_nbrs;
+	int sfc = sizeof(float) * max_nbrs;
+	int sj = sizeof(int) * max_nbrs;
+
+	return (sm_aev + sxyz + sRij + sfc + sj) * ncatom_per_tpb;
+      };
+
+      int maxNbrsPerCenterAtom = cubMax(d_numPairsPerCenterAtom, ncenter_atoms, d_count_out);
+
+      int maxnbrs_per_atom_aligned = align<4>(maxNbrsPerCenterAtom);
+
+      cuAngularAEVs
+	<<<nblocks_angAEV, block_size, smem_size(maxnbrs_per_atom_aligned, block_size/nthreads_per_catom) >>>

```

Comment:
I think we should run this on stream `c10::cuda::getCurrentCUDAStream()`, instead of the default stream.

Reply:
Working at https://github.com/akkamesh/torchani/pull/1